src/FSharpPlus/Memoization.fs

```diff
@@ -0,0 +1,17 @@
+﻿namespace FSharpPlus
+
+
+[<AutoOpen>]
+module Memoization =
+
+    // From https://gist.github.com/gusty/70e5af737f2f6aed2bc0303a2e17c7d7
+
+    open System.Collections.Concurrent
+  
+    type T = T with static member getOrAdd (cd:ConcurrentDictionary<_,'t>) (f:_ -> _) k = cd.GetOrAdd (k, f)
+  
+    let inline memoize (f:'``(T1 -> T2 -> ... -> Tn)``): '``(T1 -> T2 -> ... -> Tn)`` = (T $ Unchecked.defaultof<'``(T1 -> T2 -> ... -> Tn)``>) f
```

Comment:
I think @gusty [preferred](https://github.com/fsprojects/FSharpPlus/issues/88#issuecomment-434485033) to have the default memoize function in the default namespace.

Reply:
Hence the AutoOpen, since you can't define `let` values directly in namespaces. If I have misunderstood, please clarify where you want it placed.

Reply:
Sorry, I didn't notice the `AutoOpen`. Perhaps an implementation module named internal or something in order to hide T ?

Reply:
Are you then referring to simply renaming `module Memoization` to `module Internal` and still have AutoOpen? Or moving the type to `module Internal` and keeping `memoize` in `module Memoize`?

Reply:
Would it work to have something like:
```
    module Internal =
       open System.Collections.Concurrent
  
       type T = T with static member getOrAdd (cd:ConcurrentDictionary<_,'t>) (f:_ -> _) k = cd.GetOrAdd (k, f)
       type T with     
          static member        ($) (_:obj,  _: 'a -> 'b) = T.getOrAdd (ConcurrentDictionary ())
          static member inline ($) (T, _:'t -> 'a -> 'b) = T.getOrAdd (ConcurrentDictionary ()) << (<<) 
    open Internal
    let inline memoize (f:'``(T1 -> T2 -> ... -> Tn)``): '``(T1 -> T2 -> ... -> Tn)`` = (T $ Unchecked.defaultof<'``(T1 -> T2 -> ... -> Tn)``>) f
  
```
?

Reply:
Inside the Memoization module?

Reply:
Though, that would be somewhat the same as your suggestion to give `T` an awful name

Reply:
I'll check, I have just assumed that the `T` extensions need to be defined after `memoize` since it was that way in the original sample.

Reply:
Perhaps give it the name `MemoizeConcurrentT` or something ? What does @gusty think?

Reply:
I'd perhaps go for `MemoizeConcurrentHelper`

Reply:
And it's a bit of a mouthful so you'd not expect anyone to use it unintentionally. 😄

Reply:
Well that entirely depends upon what you imagine adding to the module in the future.

One could of course extend it with more memoization, like a thread-unsafe but more efficient version using `Dictionary`. But other than memoization, I have no idea what you'd want to place here.

Reply:
So I guess this pull request should be fine to merge then once the name of `T` is changed?

Reply:
I also have another issue; currently it fails for `null` inputs (since `ConcurrentDictionary` can't have `null` keys). I'll create a single-case struct DU to wrap the key in, that seems to be the normal approach to overcome that particular limitation in other F# memoization implementations I've seen.

Reply:
Cool !
I haven't thought about the `null` case.